modules/renter/renter.go

```diff
@@ -123,36 +122,110 @@ func newRenter(cs modules.ConsensusSet, tpool modules.TransactionPool, hdb hostD
 	if tpool == nil {
 		return nil, errNilTpool
 	}
+	if hc == nil {
+		return nil, errNilContractor
+	}
 	if hdb == nil {
 		// Nil hdb currently allowed for testing purposes. :(
 		// return nil, errNilHdb
 	}
 
 	r := &Renter{
-		cs:             cs,
-		hostDB:         hdb,
-		hostContractor: hc,
-
+		newFiles: make(chan *file),
 		files:    make(map[string]*file),
 		tracking: make(map[string]trackedFile),
 
-		persistDir: persistDir,
-		mu:         sync.New(modules.SafeMutexDelay, 1),
+		workerPool: make(map[types.FileContractID]*worker),
+
+		cs:             cs,
+		hostDB:         hdb,
+		hostContractor: hc,
+		persistDir:     persistDir,
+		mu:             sync.New(modules.SafeMutexDelay, 1),
+		tg:             new(sync.ThreadGroup),
 	}
 	if err := r.initPersist(); err != nil {
 		return nil, err
 	}
 
+	// Spin up the workers for the work pool.
+	contracts := hc.Contracts()
+	for _, c := range contracts {
+		err := r.addWorker(c.ID)
+		if err != nil {
+			return nil, err
+		}
+	}
+
 	go r.threadedRepairLoop()
```

Comment:
I think I just forgot to spawn it.

The download loop is separate from the repair loop because the events that wake the download loop are different from the events that wake the repair loop.

The workers can handle both downloads and uploads being fed to them at the same time, so it's safe to have both loops running in parallel. That simplifies needing to worry about whether or not a worker is in use. The worker will block on the upload channel if there is an active download, and will block on the download channel if there is an active upload. The worker, once ready for more work, will prioritize downloads over uploads.